chronos/asyncsync.nim

```diff
@@ -198,32 +197,13 @@ proc wakeupNext(waiters: var seq[Future[void]]) {.inline.} =
   var i = 0
   while i < len(waiters):
     var waiter = waiters[i]
+    inc(i)
+
     if not(waiter.finished()):
-      let length = len(waiters) - (i + 1)
-      let offset = len(waiters) - length
-      if length > 0:
-        for k in 0..<length:
-          shallowCopy(waiters[k], waiters[k + offset])
-      waiters.setLen(length)
       waiter.complete()
       break
-    inc(i)
 
-proc removeWaiter(waiters: var seq[Future[void]],
-                  waiter: Future[void]) {.inline.} =
-  ## Safely remove ``waiter`` from list of waiters in ``waiters``. This
-  ## procedure will not raise if ``waiter`` is not in the list of waiters.
-  var index = waiters.find(waiter)
-  if index >= 0:
-    waiters.delete(index)
-
-proc removeWaiter(waiters: var Deque[Future[void]],
-                  fut: Future[void]) {.inline.} =
-  var nwaiters = initDeque[Future[void]]()
-  while len(waiters) > 0:
-    var waiter = waiters.popFirst()
-    if waiter != fut:
-      nwaiters.addFirst(waiter)
+  waiters.delete(0, i - 1)
```

Comment:
It looks like you do not like to keep an order here, what is actually important.

Reply:
`delete` keeps order

Reply:
the significant fix here is that we remove finished futures even if there is no finished future (if it was cancelled for example) - this allows the gc to recycle it earlier.